start making some chart writing tips/notes (#795)
alphagov/gsp: docs/gds-supported-platform/helm-chart-best-practices.md

# Helm Chart Best Practices

## Overview

> Guidance for authoring Helm chart's suitable for production deployments.

Reviewer: ```suggestion
Guidance for authoring Helm charts suitable for production deployments.
```

### Use the release name in resource names

> `.metadata.name` should include at least `{{ .Release.Name }}`

Reviewer: we should also probably give some advice about setting Release.Name

Author: Probably a bit too much of a bike shedy topic to give advice on really ... Release name could be used for anything from commit SHAs, release number and chart versions to more logical names like prod/staging or even just one offs like we do for "gsp" ... So I'd shy away :smile:

Reviewer: I mean, from experience, we've been bitten before by the release name being a long string that just identifies the app and the environment (which was too long) without it being a version.

For example:

```
metadata:
    name: {{ .Release.Name }}-my-app
```

This allows you to install multiple versions of your application to a single
namespace without clashing/clobbering resources with the same name.

### High Availability

> Replicas should be specified per application, rather than globaly so that they can we scaled individualy:

Reviewer: ```suggestion
Replicas should be specified per application, rather than globally so that they can we scaled individually:
```
also, some note about taking into account limitations given by application's state, such as rate limiting

```
myApp:
  replicas: 2
```

### Resources & Limits

> resources should be specified in pod templates (both requests and limits) when parameterized in values.yaml files like:

Reviewer: ```suggestion
resources should be specified in pod templates (both requests and limits) when parameterised in values.yaml files like:
```

```
  myApp:
    resources:
      requests:
        cpu: 500m
        memory: 128Mi
      limits:
        cpu: 1000m
        memory: 512Mi
```

Which can then be consumed in templates with the `toYaml` helper:

```
  spec:
    template:
      spec:
        containers:
          - name: my-app
            resources:
  {{ toYaml .Values.resources | indent 12 }}
```

### Readiness and liveness probes

Use both readiness probes and liveness probes

### Set an update strategy on deployments

```
  spec:
    strategy:
      type: RollingUpdate
      rollingUpdate:
        maxUnavailable: 25%
        maxSurge: 25%
  ```

### One resource per file

> avoid `kind: List` resources; one resource per file

Reviewer: also avoid multi-YAML-document files?
